### PR TITLE
[SMTChecker] Report SMTEncoder warnings also via CHC

### PR DESCRIPTION
**Pull request by leonardoalt**
Currently CHC does not forward the unsupported warning messages from SMTEncoder to the ErrorReporter given by CompilerStack, since it relies on BMC running after and doing that. Since now engines can be run separately, this is a problem if you run CHC only, since the "unsupported" messages will *not* be displayed.
This PR makes CHC forward the SMTEncoder messages similar to what BMC does.
The effect of that is getting duplicated messages.
I think that's fine for now because:
- We want to support the reported features and remove those messages anyway
- It's better to make it safe right away and then improve the readability of the messages

### CODE REVIEW

File: libsolidity/formal/CHC.cpp [hunk 1]
```diff
@@ -79,6 +79,26 @@ void CHC::analyze(SourceUnit const& _source)
 		source->accept(*this);
 
 	checkVerificationTargets();
+
+	bool ranSolver = true;
+#ifndef HAVE_Z3
+	auto solver = dynamic_cast<CHCSmtLib2Interface const*>(m_interface.get());
```
Comment (mijovic) [inline libsolidity/formal/CHC.cpp]:
```suggestion
	auto const* solver = dynamic_cast<CHCSmtLib2Interface const*>(m_interface.get());
```

File: libsolidity/formal/SMTEncoder.h [hunk 2]
```diff
@@ -313,8 +313,6 @@ class SMTEncoder: public ASTConstVisitor
 
 	smt::VariableUsage m_variableUsage;
 	bool m_arrayAssignmentHappened = false;
-	// True if the "No SMT solver available" warning was already created.
-	bool m_noSolverWarning = false;
```
Comment (mijovic) [inline libsolidity/formal/SMTEncoder.h]: Why did you need to move this from here?

**Review by mijovic — COMMENTED** [review]
I noticed in some of tests that BMC reports multiple unsupported warnings for same location, while CHC gives only one. I guess this is fine.

**Comment by leonardoalt** [comment]
> I noticed in some of tests that BMC reports multiple unsupported warnings for same location, while CHC gives only one. I guess this is fine.

Yea I noticed that too, also CHC sometimes gives `Internal expression not defined` whereas BMC doesn't, but I think it should be solved separately.

File: libsolidity/formal/SMTEncoder.h (hunk 2 above)
Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.h]: Because now BMC and CHC each need one separately

Comment (mijovic) [reply libsolidity/formal/SMTEncoder.h]: I get that, but making it protected in `SMTEncoder` should work?
I mean, instances of CHC and BMC will have it's own value for it. Or I am missing something

Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.h]: Right, now I don't know actually. I thought they could conflict somehow but that doesn't make sense, will move it back

File: libsolidity/formal/CHC.cpp (hunk 1 above)
Comment (leonardoalt) [reply libsolidity/formal/CHC.cpp]: Ah this was actually debug leftover, removed

**Comment by leonardoalt** [comment]
Updated

**Review by mijovic — APPROVED** [review]
If all the tests are passing, good to merge
